scio-extra/src/main/java/com/spotify/scio/extra/transforms/DoFnWithResource.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.transforms;
+
+import com.google.common.collect.Maps;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.UUID;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * A {@link DoFn} that manages an external resource.
+ */
+public abstract class DoFnWithResource<InputT, OutputT, ResourceT> extends DoFn<InputT, OutputT> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DoFnWithResource.class);
+  private static final ConcurrentMap<String, Object> resources = Maps.newConcurrentMap();
+
+  private final String instanceId;
+  private String resourceId = null;
+
+  /**
+   * Resource type for sharing the resource among {@link DoFn} instances.
+   *
+   * {@link DoFn}s are defined and created locally, serialized, submitted to the runner and
+   * de-serialized on remote workers.
+   *
+   * <pre><code>
+   * // Define a sub-class
+   * {@literal class MyDoFn extends DoFnWithResource<String, String, MyResource>} {
+   *   ...
+   * }
+   *
+   * // Create instances
+   * MyDoFn f1 = new MyDoFn();
+   * MyDoFn f2 = new MyDoFn();
+   *
+   * {@literal PCollection<String>} lines = ...;
+   *
+   * // Clone instances to remote workers
+   * // On a 4-core Dataflow worker, both `f1` and `f2` are cloned 4 times, one for each CPU core.
+   * lines.apply(ParDo.of(f1));
+   * lines.apply(ParDo.of(f2));
+   * </code></pre>
+   */
+  public enum ResourceType {
+    /**
+     * One instance of the resource per sub-class.
+     *
+     * All instances of the same class in a remote JVM share a single `ResourceT`, e.g. all clones
+     * of `f1` and `f2`. This is useful for sharing resources globally, e.g. static look up table,
```

Comment:
Would remove `globally` it's misleading - maybe be explicit and say `sharing resources with the same JVM`.